File: prototype/rails/src/app/actions/orders/order_lifecycle_test.rb
require "commerce_test_case"

# Walks the whole order lifecycle across two sellers. Every other action test
# covers one step; these two protect the way the steps add up.
module Orders
  class OrderLifecycleTest < CommerceTestCase
    def test_an_order_runs_from_the_cart_to_the_weekly_payout
      painter = seller("Blue Kiln Studio")
      printer = seller("Rye Press")
      painting = listing(painter, price_cents: 45_000, quantity: 1)
      print = listing(printer, price_cents: 12_000, quantity: 1)

      order = order_for(customer, painting, print)

      assert_equal Domain::Orders::OrderStatus::AWAITING_PAYMENT, order.status
      assert_equal 57_000, order.total_cents
      assert_equal Domain::Listings::ListingStatus::SOLD, painting.reload.status

      FinalizeOrder.new.call(order: order, card_number: APPROVED_CARD, now: moment("2026-08-20 10:00:00"))

      assert_equal Domain::Orders::OrderStatus::PAID, order.status
      assert_equal 2, Notification.where(subject: "Item sold").count
      assert_equal [40_500, 10_800], held_per_seller(painter, printer)

      painting_shipment = order.fulfillments.find_by(seller: painter)
      print_shipment = order.fulfillments.find_by(seller: printer)

      ship(painting_shipment, "USPS", "9400111899", "2026-08-21 11:00:00")
      assert_equal Domain::Orders::OrderStatus::PARTIALLY_SHIPPED, order.reload.status

      ship(print_shipment, "FedEx", "7712349", "2026-08-21 12:00:00")
      assert_equal Domain::Orders::OrderStatus::SHIPPED, order.reload.status
      assert_equal 2, Notification.where(subject: "Order shipped").count

      deliver(painting_shipment, "2026-08-22 09:00:00")
      assert_equal Domain::Orders::OrderStatus::SHIPPED, order.reload.status

      deliver(print_shipment, "2026-08-22 10:00:00")
      assert_equal Domain::Orders::OrderStatus::DELIVERED, order.reload.status
      assert_equal [40_500, 10_800], available_per_seller(painter, printer)

      payouts = Escrow::RunWeeklyPayout.new.call(as_of: moment("2026-08-24 09:00:00"))

      assert_equal 2, payouts.size
      assert_equal({ painter.id => 40_500, printer.id => 10_800 }, Payout.order(:seller_id).pluck(:seller_id, :amount_cents).to_h)
      assert_equal [0, 0], available_per_seller(painter, printer)
      assert_equal [0, 0], held_per_seller(painter, printer)
      assert_equal [40_500, 10_800], paid_out_per_seller(painter, printer)
    end

    def test_a_declined_card_returns_the_stock_and_a_retry_completes_the_order
      shop = seller
      art = listing(shop, price_cents: 45_000, quantity: 1)
      order = order_for(customer, art)

      FinalizeOrder.new.call(order: order, card_number: UNFUNDED_CARD, now: moment("2026-08-20 10:00:00"))

      assert_equal Domain::Orders::OrderStatus::PAYMENT_FAILED, order.status
      assert_equal Domain::Listings::ListingStatus::FOR_SALE, art.reload.status
      assert_equal 1, art.quantity
      assert_equal 0, LedgerEntry.count

      FinalizeOrder.new.call(order: order, card_number: APPROVED_CARD, now: moment("2026-08-20 10:05:00"))

      assert_equal Domain::Orders::OrderStatus::PAID, order.status
      assert_equal Domain::Listings::ListingStatus::SOLD, art.reload.status
      assert_equal 0, art.quantity
      assert_equal %w[declined approved], order.payments.order(:id).pluck(:status)
      assert_equal [40_500], held_per_seller(shop)

      fulfillment = ship(order.fulfillments.sole, "USPS", "9400111899", "2026-08-21 11:00:00")
      deliver(fulfillment, "2026-08-22 09:00:00")

      payouts = Escrow::RunWeeklyPayout.new.call(as_of: moment("2026-08-24 09:00:00"))

      assert_equal [40_500], payouts.map(&:amount_cents)
      assert_equal [40_500], paid_out_per_seller(shop)
    end

    private

    def ship(fulfillment, carrier, tracking_number, at)
      Fulfillments::MarkShipped.new.call(
        fulfillment: fulfillment, carrier: carrier, tracking_number: tracking_number, now: moment(at)
      )
    end

    def deliver(fulfillment, at)
      Fulfillments::ConfirmDelivered.new.call(fulfillment: fulfillment.reload, now: moment(at))
    end

    def held_per_seller(*sellers)
      sellers.map { |shop| balance_of(shop).held.cents }
    end

    def available_per_seller(*sellers)
      sellers.map { |shop| balance_of(shop).available.cents }
    end

    def paid_out_per_seller(*sellers)
      sellers.map { |shop| balance_of(shop).paid_out.cents }
    end
  end
end

File: prototype/php/src/app/Actions/Orders/OrderLifecycleTest.php
<?php

declare(strict_types=1);

namespace App\Actions\Orders;

use App\Actions\Escrow\RunWeeklyPayout;
use App\Actions\Fulfillment\ConfirmDelivered;
use App\Actions\Fulfillment\MarkShipped;
use App\Domain\Escrow\LedgerBalance;
use App\Domain\Listings\ListingStatus;
use App\Domain\Orders\OrderStatus;
use App\Domain\Payments\PaymentStatus;
use App\Models\LedgerEntry;
use App\Models\Payout;
use App\Models\Seller;
use App\Notifications\ItemSold;
use App\Notifications\OrderShipped;
use Illuminate\Notifications\DatabaseNotification;

/**
 * Walks the whole order lifecycle across two sellers. Every other action test
 * covers one step; these two protect the way the steps add up.
 */
it('runs an order from the cart to the weekly payout', function (): void {
    $balanceOf = fn (Seller $seller): LedgerBalance => LedgerBalance::from(
        LedgerEntry::query()
            ->where('seller_id', $seller->id)
            ->get()
            ->map(fn (LedgerEntry $entry) => $entry->toMovement())
            ->all(),
    );
    $heldPerSeller = fn (Seller ...$sellers): array => array_map(fn (Seller $seller): int => $balanceOf($seller)->held->cents, $sellers);
    $availablePerSeller = fn (Seller ...$sellers): array => array_map(fn (Seller $seller): int => $balanceOf($seller)->available->cents, $sellers);
    $paidOutPerSeller = fn (Seller ...$sellers): array => array_map(fn (Seller $seller): int => $balanceOf($seller)->paidOut->cents, $sellers);

    $painter = $this->seller('Blue Kiln Studio');
    $printer = $this->seller('Rye Press');
    $painting = $this->listing($painter, ['price_cents' => 45000, 'quantity' => 1]);
    $print = $this->listing($printer, ['price_cents' => 12000, 'quantity' => 1]);
    $customer = $this->verifiedCustomer();

    $order = $this->orderFor($customer, $painting, $print);

    expect($order->status)->toBe(OrderStatus::AwaitingPayment)
        ->and($order->total_cents)->toBe(57000);
    expect($painting)->toHaveStatus(ListingStatus::Sold);

    $order = app(FinalizeOrder::class)($order, '4242 4242 4242 4242', $this->moment('2026-08-20 10:00:00'));

    expect($order->status)->toBe(OrderStatus::Paid)
        ->and(DatabaseNotification::query()->where('type', ItemSold::class)->count())->toBe(2)
        ->and($heldPerSeller($painter, $printer))->toBe([40500, 10800]);

    $paintingShipment = $order->fulfillments()->where('seller_id', $painter->id)->sole();
    $printShipment = $order->fulfillments()->where('seller_id', $printer->id)->sole();

    app(MarkShipped::class)($paintingShipment, 'USPS', '9400111899', $this->moment('2026-08-21 11:00:00'));
    expect($order)->toHaveStatus(OrderStatus::PartiallyShipped);

    app(MarkShipped::class)($printShipment, 'FedEx', '7712349', $this->moment('2026-08-21 12:00:00'));
    expect($order)->toHaveStatus(OrderStatus::Shipped);
    expect(DatabaseNotification::query()->where('type', OrderShipped::class)->count())->toBe(2);

    app(ConfirmDelivered::class)($paintingShipment->refresh(), $this->moment('2026-08-22 09:00:00'));
    expect($order)->toHaveStatus(OrderStatus::Shipped);

    app(ConfirmDelivered::class)($printShipment->refresh(), $this->moment('2026-08-22 10:00:00'));
    expect($order)->toHaveStatus(OrderStatus::Delivered);
    expect($availablePerSeller($painter, $printer))->toBe([40500, 10800]);

    $payouts = app(RunWeeklyPayout::class)($this->moment('2026-08-24 09:00:00'));

    expect($payouts)->toHaveCount(2);
    expect(Payout::query()->orderBy('seller_id')->pluck('amount_cents', 'seller_id')->all())
        ->toBe([$painter->id => 40500, $printer->id => 10800]);
    expect($availablePerSeller($painter, $printer))->toBe([0, 0])
        ->and($heldPerSeller($painter, $printer))->toBe([0, 0])
        ->and($paidOutPerSeller($painter, $printer))->toBe([40500, 10800]);
});

it('returns the stock on a declined card and completes the order on retry', function (): void {
    $balanceOf = fn (Seller $seller): LedgerBalance => LedgerBalance::from(
        LedgerEntry::query()
            ->where('seller_id', $seller->id)
            ->get()
            ->map(fn (LedgerEntry $entry) => $entry->toMovement())
            ->all(),
    );
    $heldPerSeller = fn (Seller ...$sellers): array => array_map(fn (Seller $seller): int => $balanceOf($seller)->held->cents, $sellers);
    $paidOutPerSeller = fn (Seller ...$sellers): array => array_map(fn (Seller $seller): int => $balanceOf($seller)->paidOut->cents, $sellers);

    $seller = $this->seller();
    $listing = $this->listing($seller, ['price_cents' => 45000, 'quantity' => 1]);
    $customer = $this->verifiedCustomer();
    $order = $this->orderFor($customer, $listing);
    $finalizeOrder = app(FinalizeOrder::class);

    $order = $finalizeOrder($order, '4000 0000 0000 9995', $this->moment('2026-08-20 10:00:00'));

    expect($order->status)->toBe(OrderStatus::PaymentFailed);
    expect($listing)->toHaveStatus(ListingStatus::ForSale);
    expect($listing->fresh()->quantity)->toBe(1)
        ->and(LedgerEntry::query()->count())->toBe(0);

    $order = $finalizeOrder($order, '4242 4242 4242 4242', $this->moment('2026-08-20 10:05:00'));

    expect($order->status)->toBe(OrderStatus::Paid);
    expect($listing)->toHaveStatus(ListingStatus::Sold);
    expect($listing->fresh()->quantity)->toBe(0)
        ->and($order->payments()->orderBy('id')->pluck('status')->all())->toBe([PaymentStatus::Declined, PaymentStatus::Approved])
        ->and($heldPerSeller($seller))->toBe([40500]);

    $fulfillment = app(MarkShipped::class)($order->fulfillments()->sole(), 'USPS', '9400111899', $this->moment('2026-08-21 11:00:00'));
    app(ConfirmDelivered::class)($fulfillment, $this->moment('2026-08-22 09:00:00'));

    $payouts = app(RunWeeklyPayout::class)($this->moment('2026-08-24 09:00:00'));

    expect(array_map(fn (Payout $payout): int => $payout->amount_cents, $payouts))->toBe([40500])
        ->and($paidOutPerSeller($seller))->toBe([40500]);
});
